### DISCUSSION TITLE
Bluepill support usb bootloader

### DISCUSSION THREAD

**Discussion by yashikada** [blackmagic-debug/blackmagic#1357] [Q&A] [answered]
I have a bluepill with 128KB of flash so big enough for the current version of BMP

```
Memory region         Used Size  Region Size  %age Used
             rom:      114656 B       128 KB     87.48%
             ram:       11584 B        20 KB     56.56%
```

Can I use usb bootloader [STM32duino-bootloader](https://github.com/rogerclarkmelbourne/STM32duino-bootloader) ?

I tried to add this to src/Makefile

```
ifneq ($(FIRMWARE_ORIGIN),)
LDFLAGS         += -Wl,-section-start=.isr_vector=$(FIRMWARE_ORIGIN)
endif
```
and compile with `make PROBE_HOST=swlink FIRMWARE_ORIGIN=0x08002000` but doesn't works.

**Comment by dragonmux** [comment, marked as answer]
So.. two things:

1) in theory, having taken a quick look at that bootloader, yes, it should work (though we provide a DFU bootloader, so.. why?)
2) the platform determines the section start point for the firmware, your additions to src/Makefile are in conflict with the existing ones for the platform (also, the interrupt vector handlers aren't in a section called .isr_vector in the BMP firmware). See the platform Makefile chunk below:

https://github.com/blackmagic-debug/blackmagic/blob/98915144944d46cb2dc1daedc734b0134dbde7fb/src/platforms/swlink/Makefile.inc#L12

It would be useful if you could share what you're doing to try to write new firmware to your board.
Another thing to note is that for Bluepill, you want to use the stlink platform, not the swlink platform - compile with `make PROBE_HOST=stlink BLUEPILL=1`

**Reply by yashikada** [reply]
>     1. in theory, having taken a quick look at that bootloader, yes, it should work (though we provide a DFU bootloader, so.. why?)

only because the bootloader is already written in my bluepill.
 
>     2. the platform determines the section start point for the firmware, your additions to src/Makefile are in conflict with the existing ones for the platform (also, the interrupt vector handlers aren't in a section called .isr_vector in the BMP firmware). See the platform Makefile chunk below:
> 
> 
> https://github.com/blackmagic-debug/blackmagic/blob/98915144944d46cb2dc1daedc734b0134dbde7fb/src/platforms/swlink/Makefile.inc#L12
> 

ok it's work without my change.

> It would be useful if you could share what you're doing to try to write new firmware to your board. Another thing to note is that for Bluepill, you want to use the stlink platform, not the swlink platform - compile with `make PROBE_HOST=stlink BLUEPILL=1`

for now i just wanted to test the firmware and figure out how it worked, no firmware to write

**Reply by dragonmux** [reply]
> 
> only because the bootloader is already written in my bluepill.
> 
Ah, fair enough! good reason
> 
> > It would be useful if you could share what you're doing to try to write new firmware to your board. Another thing to note is that for Bluepill, you want to use the stlink platform, not the swlink platform - compile with `make PROBE_HOST=stlink BLUEPILL=1`
> 
> for now i just wanted to test the firmware and figure out how it worked, no firmware to write
>
That we understand.. we'll re-word the question differently: You tried putting firmware on your Bluepill, we are asking how you did this - what command you ran, please?

You've not given us any information on how you got from a file on disk from the build system to firmware on the Bluepill, which we suspect is where your troubles lie.

For example, if you used blackmagic_dfu.bin then that will not work (that is the bootloader). If you used blackmagic.elf with dfu-util that also will not work because it needs a raw image. You specifically need to use blackmagic.bin if using dfu-util, but other utilities are smarter and allow you to use the .elf instead

**Reply by yashikada** [reply]
ok sorry If I don't understand very well the question.
I used this command: `dfu-util -a 2 blackmagic.bin`

After you reply I changed all. I replaced the previous bootloader with blackmagic_dfu.bin and I can run the firmware without issue, with other bootloader the firmware start only one time after the reset.
Now no issue, it's all rigth.

**Reply by dragonmux** [reply]
We're glad to hear you got it working.

If we find some time in the next week or so we'll look into that other bootloader and see why using its second slot didn't work reliably, though fundamentally that sounds like a bug in that bootloader so if you wish to pursue using it, you may wish to file a bug on that project referencing the observed behaviour.

The Black Magic Firmware should at least boot reliably even if it can't get back into the bootloader due to potentially talking a different bootloader reboot protocol.
